### PR TITLE
Add functions to compute interface projection operators

### PR DESCRIPTION
**Pull request by prayush**
## Proposed changes

Add functions to compute operators to project spatial tensors onto an interface.

### Code review checklist

- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).
- [ ] The PR lists upgrade instructions and is labeled `bugfix` or
  `major new feature` if appropriate.

### CODE REVIEW

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: you only need `TypeAliases.hpp` in the header file, then include `Tensor.hpp` in the `cpp` file. Type aliases basically is a bunch of forward declarations :)

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: No `gamma^{ij}` in this one :)

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp]: include `Utilities/ContainerHelpers.hpp` and use `destructive_resize_components(projection_tensor, get<0>(normal_vector).size())` If `destructive_resize_components` doesn't work on `double`, then just wrap in an `if constexpr (not std::is_same_v<double, DataType>)`

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp]: same comment about trying `destructive_resize_components`

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp:74]: It's not clear to me what this test function is adding over the python tests. Rather than comparing to a python implementation you seem to be comparing to values you worked out by hand or elsewhere? At least a short comment at the start of the function describing what the test is doing would be helpful

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp
Comment (nilsdeppe) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp]: `BC action` -> `projection action`?

**Review by nilsdeppe — CHANGES_REQUESTED** [review]

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp]: corrected

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp:74]: This test is only comparing the output of these functions to values from spec. I've added comments to explain this.

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp]: updated

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp]: done.

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: corrected help text

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: thanks for catching this, fixed.

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (nilsdeppe) [inline src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: you lost the `_` to make the indices downstairs on this line and the line above

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please squash, including this minor fix

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: [very nitpick] the spacing before and after the \f$ is different for each expression in the dox here.

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp
Comment (moxcodes) [inline src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp]: Initialize these with their size to avoid forcing the destructive resize (those have an `UNLIKELY`, so it's better to have the 'typical' case initialize the size when you know it will be necessary to do so).

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp
Comment (moxcodes) [inline tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp]: This only iterates over the upper-triangular matrix elements, and because it is initialized with the expected object before the assignment, the missing assignments aren't obvious.
I suggest initializing the `spec_spatial_projection...` to the default constructed type (or `make_with_values` if you prefer), so that it's more obvious that the test and expected are independently constructed.

**Review by moxcodes — COMMENTED** [review]
Just a few small suggestions, overall looks good

File: tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp
Comment (prayush) [reply tests/Unit/PointwiseFunctions/GeneralRelativity/Test_ProjectionOperators.cpp]: makes sense, changed to initializing with `make_with_value` everywhere.

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.cpp]: done.

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: fixed!

File: src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp
Comment (prayush) [reply src/PointwiseFunctions/GeneralRelativity/ProjectionOperators.hpp]: argh, fixed!

**Review by nilsdeppe — COMMENTED** [review]
LGTM, please go ahead and squash once @moxcodes has had a chance to look at the fixup

**Comment by moxcodes** [comment]
I'm happy with the fixup. Go ahead and squash.

**Comment by prayush** [comment]
@sxs-collaboration/spectre-core-devs - this is good to go.

**Review by nilsdeppe — APPROVED** [review]

**Comment by wthrowe** [comment]
I can't merge this until it has been approved by the reviewers...

**Review by moxcodes — APPROVED** [review]

**Pull request merged**
